src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.cpp

```diff
@@ -0,0 +1,764 @@
+
+	_performance_model.updateParameters();
+	_roll_slew_rate.setSlewRate(radians(_param_fw_pn_r_slew_max.get()));
+
+	_tecs_alt_time_const_slew_rate.setSlewRate(TECS_ALT_TIME_CONST_SLEW_RATE);
+	_tecs_alt_time_const_slew_rate.setForcedValue(_param_fw_t_h_error_tc.get() * _param_fw_thrtc_sc.get());
+}
+
+void FwLateralLongitudinalControl::Run()
+{
+	if (should_exit()) {
+		_local_pos_sub.unregisterCallback();
+		exit_and_cleanup();
+		return;
+	}
+
+	perf_begin(_loop_perf);
+
+	/* only run controller if position changed */
+	// check for parameter updates
+	if (_parameter_update_sub.updated()) {
+		// clear update
+		parameter_update_s pupdate;
+		_parameter_update_sub.copy(&pupdate);
+
+		// update parameters from storage
+		parameters_update();
+	}
+
+	if (_local_pos_sub.update(&_local_pos)) {
+
+		const float control_interval = math::constrain((_local_pos.timestamp - _last_time_loop_ran) * 1e-6f,
+					       0.001f, 0.1f);
+		_last_time_loop_ran = _local_pos.timestamp;
+
+		updateControlLimits();
+
+		_tecs.set_speed_weight(_long_limits.speed_weight);
+		updateTECSAltitudeTimeConstant(checkLowHeightConditions()
+					       || _long_limits.enforce_low_height_condition, control_interval);
+		_tecs.set_altitude_error_time_constant(_tecs_alt_time_const_slew_rate.getState());
+
+		if (_vehicle_air_data_sub.updated()) {
+			_air_density = PX4_ISFINITE(_vehicle_air_data_sub.get().rho) ? _vehicle_air_data_sub.get().rho : _air_density;
+			_tecs.set_max_climb_rate(_performance_model.getMaximumClimbRate(_air_density));
+			_tecs.set_min_sink_rate(_performance_model.getMinimumSinkRate(_air_density));
+		}
+
+		if (_vehicle_landed_sub.updated()) {
+			vehicle_land_detected_s landed{};
+			_vehicle_landed_sub.copy(&landed);
+			_landed = landed.landed;
+		}
+
+		_flight_phase_estimation_pub.get().flight_phase = flight_phase_estimation_s::FLIGHT_PHASE_UNKNOWN;
+
+		_vehicle_status_sub.update();
+		_control_mode_sub.update();
+		update_control_state();
+
+		if (_control_mode_sub.get().flag_control_manual_enabled && _control_mode_sub.get().flag_control_altitude_enabled
+		    && _local_pos.z_reset_counter != _z_reset_counter) {
+			if (_control_mode_sub.get().flag_control_altitude_enabled && _local_pos.z_reset_counter != _z_reset_counter) {
+				// make TECS accept step in altitude and demanded altitude
+				_tecs.handle_alt_step(_long_control_state.altitude_msl, _long_control_state.height_rate);
+			}
+		}
+
+		const bool should_run = _control_mode_sub.get().flag_control_position_enabled ||
+					_control_mode_sub.get().flag_control_velocity_enabled ||
+					_control_mode_sub.get().flag_control_acceleration_enabled ||
+					_control_mode_sub.get().flag_control_altitude_enabled ||
+					_control_mode_sub.get().flag_control_climb_rate_enabled;
+
+		if (should_run) {
+			float pitch_sp{NAN};
+			float throttle_sp{NAN};
+
+			if (_fw_longitudinal_ctrl_sub.updated()) {
+				_fw_longitudinal_ctrl_sub.copy(&_long_control_sp);
```

Comment:
```suggestion
                                fw_longitudinal_control_setpoint_s long_control_sp
				_fw_longitudinal_ctrl_sub.copy(&long_control_sp);
```
Looks like it's only in this scope so definition can be within the scope instead of the headerfile

Reply:
yeah but `_fw_longitudinal_ctrl_sub.updated()` may not always be true, and in these cases we still want to execute the code here but with the value form a previous update.